File: problem-solving/lc-70-easy-climbing-stairs/1-top-down.go
/**
 * https://leetcode.com/problems/climbing-stairs/
 * 
 * Complexity
 * ----------
 * Time: O(N)
 * Space: O(N)
 */

var mem = make(map[int]int)

func climbStairs(n int) int {
    if n < 4 {
        return n
    }
    if v, ok := mem[n]; ok {
        return v
    }
    mem[n] = climbStairs(n - 1) + climbStairs(n - 2)
    return mem[n]
}

File: problem-solving/lc-70-easy-climbing-stairs/1-top-down.py
"""
https://leetcode.com/problems/climbing-stairs/

Complexity
----------
Time: O(N)
Space: O(N)
"""

from functools import lru_cache

class Solution:
    @lru_cache(None)
    def climbStairs(self, n: int) -> int:
        if n < 3:
            return n
        return self.climbStairs(n - 1) + self.climbStairs(n - 2)
